keps/sig-storage/4650-stateful-set-update-claim-template/README.md

```diff
@@ -0,0 +1,1023 @@
+* Automatically update the associated PersistentVolumeClaim objects in-place if applicable.
+* Support updating PersistentVolumeClaim objects with `OnDelete` strategy.
+* Coordinate updates to `Pod` and PersistentVolumeClaim objects.
+* Provide accurate status and error messages to users when the update fails.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+* Support automatic rolling update of PersistentVolumeClaim.
+* Validate the updated `volumeClaimTemplate` as how PVC update does.
+* Update ephemeral volumes.
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+1. Change API server to allow any updates to `volumeClaimTemplate` of a StatefulSet.
+
+2. Modify StatefulSet controller to add PVC reconciliation logic.
+
+3. Collect the status of managed PVCs, and show them in the StatefulSet status.
+
+### Kubernetes API Changes
+
+Changes to StatefulSet `spec`:
+
+1. Introduce a new field in StatefulSet `spec`: `volumeClaimUpdateStrategy` to
+   specify how to coordinate the update of PVCs and Pods. Possible values are:
+   - `OnDelete`: the default value, only update the PVC when the the old PVC is deleted.
+   - `InPlace`: update the PVC in-place if possible. Also includes the `OnDelete` behavior.
+
+2. Introduce a new field in StatefulSet `spec.updateStrategy.rollingUpdate`: `volumeClaimSyncStrategy`
+   to specify how to update PVCs and Pods. Possible values are:
+   - `Async`: the default value, preseve the current behavior.
+   - `LockStep`: update PVCs first, then update Pods. See below for details.
+
+Changes to StatefultSet `status`:
+
+Additionally collect the status of managed PVCs, and show them in the StatefulSet status.
+
+For each PVC in the template:
+- compatible: the number of PVCs that are compatible with the template.
+  These replicas will not be blocked on Pod recreation if `volumeClaimSyncStrategy` is `LockStep`.
+- updating: the number of PVCs that are being updated in-place.
+- overSized: the number of PVCs that are over-sized.
+- totalCapacity: the sum of `status.capacity` of all the PVCs.
+
+Some fields in the `status` are also updated to reflect the staus of the PVCs:
+- readyReplicas: in addition to pods, also consider the PVCs status. A PVC is not ready if:
+  - `volumeClaimUpdateStrategy` is `InPlace` and the PVC is updating;
+- availableReplicas: total number of replicas of which both Pod and PVCs are ready for at least `minReadySeconds`
+- currentRevision, updateRevision, currentReplicas, updatedReplicas
+  are updated to reflect the status of PVCs.
```

Comment:
When expanding the size of pvcs and then kubectl rollout undo, actually pvcs will not be rollbacked, they will stay the expanded size. Right?

Reply:
It depends. If you enabled feature gate RecoverVolumeExpansionFailure, and the PVC `pvc.Status.AllocatedResources` is lower, then PVC is rolled back, and expansion is cancelled.

This is what I mean by writing "update the PVC in-place to the extent possible". Maybe I should make this more explicit.